test/dhtest.c

```diff
@@ -62,11 +62,15 @@ static int dh_test(void)
         || !TEST_true(DH_set0_pqg(dh, p, q, g)))
         goto err;
 
+    p1=p, p=NULL;
+    q1=q, q=NULL;
+    g1=g, g=NULL;
```

Comment:
Does this kind of expression adhere to the code style guidelines?

Reply:
(applies to the above 3 lines)

Reply:
I'd also like if there was a comment pointing at `DH_set0_pgq` (and the related `BN_free`s in case of early error) as the motivation of `p = q = g = NULL`, while we want to save the addresses for the subsequent tests.

Reply:
As an alternative the problem with `p`, `q` and `g` could be solved by replacing the first five `goto err`  with `goto early_err` and then adding at the end of the function:

```diff
     return ret;
+
+ early_err:
+    DH_free(dh);
+    BN_free(p);
+    BN_free(q);
+    BN_free(g);
+    BN_free(priv_key);
+    return 0;
 }
```
and omitting `BN_free()` on `p`, `q`, `g` and `priv_key` in the *normal* `err` handling.

This should allow to avoid `p1`, `q1`, `g1` and `priv_key1` at all.

Reply:
Thanks for the suggestions. I'll think about them and push an update it as soon as possible.

Reply:
Even better, if we moved lines 83 and 84 so they are part of the if starting at line 59, we could limit the `early_err` to only the first two `goto err`, avoiding a few corner cases.

What I am proposing basically wold be to group together:

```c
if (!TEST_true(BN_set_word(p, 4079L))
         || !TEST_true(BN_set_word(q, 2039L))
         || !TEST_true(BN_set_word(g, 3L))
         || !TEST_true(BN_set_word(priv_key, 1234L))
         || !TEST_true(DH_set0_pqg(dh, p, q, g))
         || !TEST_true(DH_set0_key(dh, NULL, priv_key)))
    goto early_err;
```

Reply:
Well, not quite. But I get your point. There are three error cases, see the new fixup I pushed.